Introduce script coding standards (#51)
Mbed-TLS/mbedtls-docs: kb/development/script-coding-standards.md

# Mbed TLS script coding standards

## Choice of scripting language

The preferred language for scripts in the Mbed TLS project is Python. Use Python for new scripts that are committed to the project unless there's a good reason.

Some reasons not to use Python:

* A script that only applies to Windows and should ideally work out of the box can be written in cmd.
* Some CI code in the `mbedtls-test` repository is written in Groovy, for its Jenkins integration.
* A script only intended for the Mbed TLS developers, and whose job is mostly to run other programs, can be written in sh.
* If you're extending existing scripts, then obviously keep using the same language.
* Throwaway scripts that do ad hoc automated refactoring (e.g. renaming identifiers) are often in Perl, thanks to its convenient text manipulation primitives.

> ### Language versions

Reviewer: So the file is "script-coding-standards", and it (very reasonably) starts with a section on choice of language, and (very reasonably) has a section on python coding standards, but in between there are sections on language version, which is only about python, and naming conventions, which is about all types of scripts, and invocation, which seems to only be about Unix systems (do we expect scripts written for `cmd.exe` to take `--help` or `/help`?

Suggestion: move the "Language versions" section into the Python-specific part of this document

Check [`README.md`](https://github.com/Mbed-TLS/mbedtls/blob/development/README.md#tool-versions) for the current minimum supported version of Python and other tools. Note that in addition to the public commitments, at any given time, there may be additional constraints due to the platforms on which we run automated tests. As of September 2022, Python scripts must be compatible with Python 3.5.

> For scripts that need to be backported, note that [long-time support branches](https://github.com/Mbed-TLS/mbedtls/blob/development/BRANCHES.md#long-time-support-branches) may support older versions of Python and other tools than the `development` branch.

Reviewer 2: Type annotations were added to python in version 3.5, so those scripts wouldn't run on anything older than that.
(Comment-style annotations are compatible with earlier python versions as well.)

I think we implicitly dropped support for older version of python in the development (2.x) branch when we stopped testing on Ubuntu 14.04, which ended LTS support in 2019.

> For scripts that need to be backported, note that [long-time support branches](https://github.com/Mbed-TLS/mbedtls/blob/development/BRANCHES.md#long-time-support-branches) may support older versions of Python and other tools than the `development` branch.

Author: `config.py`, which is the one script that ordinary users need, still appears to work under Python 3.4. It's from before we added type annotations. Many other scripts require at least Python 3.5 just to parse the type annotations.

We never had CI with Python 3.4: when we switched from Python 2 to Python 3, our CI was Ubuntu 16.04 only (plus more recent Python on FreeBSD and Windows).

I think it would make sense to preserve backward compatibility with Python 3.4 for `config.py` because it's essential. The other scripts are for testing or for contributors, and we can require a more recent version even in the LTS branch.

I think it's better if there's a single point of truth for minimal versions, in individual branches's README, rather than duplicate the information in the coding standards and predictably forget to update it. The only reason I mentioned 3.5 here is that the requirement comes from the current CI, it's not something you can deduce from reading the code in the mbedtls repository.

> For scripts that need to be backported, note that [long-time support branches](https://github.com/Mbed-TLS/mbedtls/blob/development/BRANCHES.md#long-time-support-branches) may support older versions of Python and other tools than the `development` branch.

Reviewer 3: Agree about specifying versions in the branches themselves (which we have started doing for development and perhaps should add to 2.28). If config.py is the solitary exception, then we should note that in a comment at the top of config.py.

> For scripts that need to be backported, note that [long-time support branches](https://github.com/Mbed-TLS/mbedtls/blob/development/BRANCHES.md#long-time-support-branches) may support older versions of Python and other tools than the `development` branch.

Author: I've reorganized and reformulated the Python version requirements. I added an explicit mention of 3.4 for `config.py` here: it is going to remain stable for another couple of years. I'll make a PR on development and 2.28 to also mention it in `config.py` itself.

> ### Script naming conventions

Reviewer 2: Nit: I don't think "Script naming conventions", "Invocation conventions" and "File structure" should be subsections of "Choice of scripting language", logically.

I'd prefer to have all four of them on the same level of heading.

* Use lowercase names.
* Use either underscores or hyphens to separate words. Python modules that are imported by other scripts cannot use hyphens.
* Perl scripts have a `.pl` extension. Python scripts have a `.py` extensions. Sh and bash scripts have a `.sh` extension. This helps on Windows.

### Invocation conventions

Scripts should be able to run from the root of the Mbed TLS source tree. Support for running from a subdirectory is a plus.

Scripts should run out of the box on a typical Unix-like system if possible. If the script requires some environment variables, try to set sensible defaults.

> All scripts should start with a comment that explains how to run the script. Printing those explanations when the script is invoked with no argument or with `--help` is a plus.

Author: I made `--help` a should: it's not available in all existing scripts and might be annoying if you want to use a basic `getopt`. I removed the “no argument” bit. I didn't make it a requirement for the wrong number of arguments because that's not common (utilities tend to just say “wrong usage, use `-h` for help”).

I added a guideline to use Unix-style options.

### File structure

All executable scripts committed into the repository should have the following structure:

* The first line is a [shebang line](https://en.wikipedia.org/wiki/Shebang_(Unix)), e.g. `#!/usr/bin/env python3`.
* A brief description of the file (either as a comment or as a string).

> * Copyright notice and license indication.

Reviewer 4: Should we request that the copyright information is on the first page of the document? 

In Python with extended docstrings descibing the script that can be pushed way too low.

A way of getting around it is containing the information of the docstring in the `__doc__ = """ Some Text """` format

Author: No, please. The file description goes before the license.

> * Copyright notice and license indication.

Reviewer 3: I thought we did, but I can't find this. I think we should add this to the CONTRIBUTING.md document.

> * Copyright notice and license indication.

Reviewer 3: I've created a PR for this: https://github.com/Mbed-TLS/mbedtls/pull/6284

> * Copyright notice and license indication.

Author: I just copied what was in the C coding standards. I don't think we need the details here (and risk forgetting to update them): you can just copy from an existing file.

* Initialization code.
* Auxiliary functions (unless the script is so simple it doesn't need any).
* Code that performs the script's purpose (this can be as simple as calling a `main` function, or more complex).

## Python coding standards

### Directory structure

Non-executable Python modules are in `scripts/mbedtls_dev`.
Executable Python scripts in `scripts` can use auxiliary modules with `from mbedtls_dev import …`.

Executable Python scripts in `tests/scripts` cannot refer to modules in `scripts/mbedtls_dev` directly, because we do not assume that `PYTHONPATH` contains the Mbed TLS source tree. To use other Python modules from a script in `tests/scripts`, use the following idiom:

```python
import scripts_path # pylint: disable=unused-import

> from mbedtls_dev import …

Author: :cry: I like using `…` rather than `...` because `...` looks like you're supposed to type three dots as part of the language syntax. Ah well.

```

### Python imports

> We follow the usual convention that imports go at the top of a file (after the docstring and the license comment). Imports are typically ordered as follows:

Reviewer: The "File structure" section is in the language-neutral part of this document, so should we instead make that early section language neutral and have a separate "File structure" section (with more detail) in the Python-specific part?

> We follow the usual convention that imports go at the top of a file (after the docstring and the license comment). Imports are typically ordered as follows:

Author: Imports are only tangentially related to the file structure, the discussion is big enough for a subsection of its own, and there isn't anything else to say about file structure. So I don't see the point in creating a section about python file structure.

> We follow the usual convention that imports go at the top of a file (after the docstring and the license comment). Imports are typically ordered as follows:

Reviewer 5: I was thinking of "import" as a language-neutral term - does this guidance also apply to Bash `source` and Perl `use`?

> We follow the usual convention that imports go at the top of a file (after the docstring and the license comment). Imports are typically ordered as follows:

Author: Not really applicable since we don't write substantial new sh or perl code and we aren't going to create sh or perl libraries.

1. Standard library, in lexicographic order.
2. Third-party libraries, if any, in lexicographic order.

> 3. Modules from `mbedtls_dev`, in lexicographic order.

Author: PEP8 is already referenced above.

> good to sort imported components by len

??? I've never heard of this convention and it seems utterly bizarre to me.

Avoid importing individual names from modules, as it can be difficult to figure out which module an unqualified name comes from. It's ok to import names if they're sufficiently recognizable, for example:

* Importing individual modules from `mbedtls_dev` is fine.
* Importing capitalized type names from `typing` is fine.
* Importing individual submodules from Cryptodome is fine.

### Third-party Python libraries

Avoid bringing in new dependencies, because each dependency adds risk to the project. Only use dependencies if they have a clear advantage.

The minimum supported version of third-party Python libraries is recorded in `*.requirements.txt` files in the [`scripts` directory](https://github.com/Mbed-TLS/mbedtls/tree/development/scripts). As with the [Python version](#language-versions), keep in mind that long-time support branches may need to support older versions.

### Python language features

> The maintainers of Mbed TLS are primarily C programmers, not Python programmers. It's ok to use fancy language features if they're useful, but don't necessarily assume that the reader is familiar with them. Make sure that if a reader is unfamiliar with an advanced feature, they'll be able to easily look it up.

Reviewer: ```suggestion
The maintainers of Mbed TLS are primarily C programmers, not Python programmers. It's ok to use more advanced language features if they make the code significantly simpler, but don't necessarily assume that the reader is familiar with them. Make sure that if a reader is unfamiliar with such a feature, they'll be able to easily look it up.
```

Author: Reworded, partly following your suggestions, but I kept “fancy” because I think it has the right level of imprecision here.

> * Classes, methods and simple inheritance can be used freely.

Reviewer 3: We currently support Python 3.6, so I don't think we can use this feature in even in development.

> * Classes, methods and simple inheritance can be used freely.

Author: Since the time we said we support Python 3.6, it's gone out of support. So we could switch to 3.7 as the minimum version in the development branch (and for nonessential backported scripts).

Data classes do save a minor annoyance (sometimes you just need a struct), and they aren't difficult to understand, so when they are permitted by language version considerations, I don't see why we should forbid them.

> * Avoid metaclasses and nontrivial dynamic class creation. If you use these features, write documentation that's accessible to non-experts.

Reviewer 3: Maybe strengthen to something like "Do not use metaclasses and nontrivial dynamic class creation, unless there is a very compelling reason"?

Author: I'm not sure the existing use of metaclasses, which has just been added, passes the “very compelling” bar.

* Do define [special methods](https://docs.python.org/3/reference/datamodel.html#special-method-names) if convenient for comparison, context managers, etc. But avoid customizing attribute access.
* Use [decorators](https://docs.python.org/3/glossary.html#term-decorator) if they have a clear benefit.
* Do use [generators](https://docs.python.org/3/glossary.html#term-generator) and [list comprehension](https://docs.python.org/3/glossary.html#term-list-comprehension) where warranted.

### Python typing

> Using type annotations is encouraged. We run [mypy](http://mypy-lang.org/) to verify static typing.

Reviewer 3: Can these be backported to 2.28 (see above comment about Python versions)?

Author: We've been running mypy on the CI since 2.26.

Reviewer 5: I believe there's a minor dragon here to be aware of. To represent a container that contains a specific type, you need `List`, `Tuple` and friends from `typing`:
```python
from typing import List

def f(x: List[str]):
    ...
```
However in Python 3.9, support for type-hinting in this way with builtin container types [was added](https://docs.python.org/3/whatsnew/3.9.html#type-hinting-generics-in-standard-collections), so in Python >= 3.9, you can do:
```python
def f(x: list[str]):
    ...
```
I'm not sure if it's worth documenting this separately though, I think it's covered by the general command to "be compatible with Python 3.5".

Avoid complex overloading. It's generally easier to understand code if it only expects a specific type of data.

### Python style conventions

> We generally follow [PEP 8](https://peps.python.org/pep-0008/), and [PEP 257](https://peps.python.org/pep-0257/) for docstrings.

Reviewer 4: As mentioned above, we could refer to pep for anything not covered not only docstrings.

Author: Uh, that's exactly what this sentence does?

Reviewer 5: I think this is one of those cases where it would be nice if English grammar had parentheses. What is meant is "we generally follow (PEP 8) and (PEP 257 for docstrings)", but could be read as "we generally follow (PEP 8 and PEP 257) for docstrings".

This could perhaps be reworded to "we generally follow PEP 8 for code and PEP 257 for docstrings" for the avoidance of doubt, but it seems a minor thing.

Author: Oh, I see. The comma actually makes it unambiguous, but it's subtle. I've reformulated.
